#### What is in `NASA_access_log_Jul95`?

In [1]:
head NASA_access_log_Jul95

199.72.81.55 - - [01/Jul/1995:00:00:01 -0400] "GET /history/apollo/ HTTP/1.0" 200 6245
unicomp6.unicomp.net - - [01/Jul/1995:00:00:06 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
199.120.110.21 - - [01/Jul/1995:00:00:09 -0400] "GET /shuttle/missions/sts-73/mission-sts-73.html HTTP/1.0" 200 4085
burger.letters.com - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/countdown/liftoff.html HTTP/1.0" 304 0
199.120.110.21 - - [01/Jul/1995:00:00:11 -0400] "GET /shuttle/missions/sts-73/sts-73-patch-small.gif HTTP/1.0" 200 4179
burger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /images/NASA-logosmall.gif HTTP/1.0" 304 0
burger.letters.com - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/video/livevideo.gif HTTP/1.0" 200 0
205.212.115.106 - - [01/Jul/1995:00:00:12 -0400] "GET /shuttle/countdown/countdown.html HTTP/1.0" 200 3985
d104.aa.net - - [01/Jul/1995:00:00:13 -0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
129.94.144.152 - - [01/Jul/1995:00:00:13 -0400] "GET / HTTP/1.0" 

#### Any `404`s?

In [2]:
grep 404 NASA_access_log_Jul95 | head

dd15-062.compuserve.com - - [01/Jul/1995:00:01:12 -0400] "GET /news/sci.space.shuttle/archive/sci-space-shuttle-22-apr-1995-40.txt HTTP/1.0" 404 -
ad03-031.compuserve.com - - [01/Jul/1995:00:04:40 -0400] "GET /facilities/vab.html HTTP/1.0" 200 4045
netport-27.iu.net - - [01/Jul/1995:00:10:19 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
netport-27.iu.net - - [01/Jul/1995:00:10:28 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
blv-pm0-ip28.halcyon.com - - [01/Jul/1995:00:14:17 -0400] "GET /persons/astronauts/i-to-l/lousmaJR.txt HTTP/1.0" 404 -
blv-pm0-ip28.halcyon.com - - [01/Jul/1995:00:14:32 -0400] "GET /persons/astronauts/a-to-d/beanAL.txt HTTP/1.0" 404 -
ix-tam1-26.ix.netcom.com - - [01/Jul/1995:00:16:04 -0400] "GET /shuttle/missions/sts-68/ksc-srl-image.html HTTP/1.0" 200 1404
www-a1.proxy.aol.com - - [01/Jul/1995:00:18:27 -0400] "GET /facilities/vab.html HTTP/1.0" 200 4045
cu-dialup-1005.cit.cornell.edu - - [01/Jul/1995:00:18:39 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 4

#### Notice a false positive on line 2:  

    ad03-031.compuserve.com - - [01/Jul/1995:00:04:40 -0400] "GET /facilities/vab.html HTTP/1.0" 200 4045 
                                                                                                     ^^^^ 

In [3]:
grep ' 404 ' NASA_access_log_Jul95 | head

dd15-062.compuserve.com - - [01/Jul/1995:00:01:12 -0400] "GET /news/sci.space.shuttle/archive/sci-space-shuttle-22-apr-1995-40.txt HTTP/1.0" 404 -
netport-27.iu.net - - [01/Jul/1995:00:10:19 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
netport-27.iu.net - - [01/Jul/1995:00:10:28 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
blv-pm0-ip28.halcyon.com - - [01/Jul/1995:00:14:17 -0400] "GET /persons/astronauts/i-to-l/lousmaJR.txt HTTP/1.0" 404 -
blv-pm0-ip28.halcyon.com - - [01/Jul/1995:00:14:32 -0400] "GET /persons/astronauts/a-to-d/beanAL.txt HTTP/1.0" 404 -
cu-dialup-1005.cit.cornell.edu - - [01/Jul/1995:00:18:39 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
cu-dialup-1005.cit.cornell.edu - - [01/Jul/1995:00:18:45 -0400] "GET /pub/winvn/readme.txt HTTP/1.0" 404 -
cu-dialup-1005.cit.cornell.edu - - [01/Jul/1995:00:18:49 -0400] "GET /pub/winvn/release.txt HTTP/1.0" 404 -
mimas.execpc.com - - [01/Jul/1995:00:18:59 -0400] "GET /shuttle/missions/technology/sts-newsref/stsref-toc.

Better, but still not fool proof. Maybe we can `cut` out just the status code.  

    199.72.81.55 - - [01/Jul/1995:00:00:01 -0400] "GET /history/apollo/ HTTP/1.0" 200 6245
    1             2 3                       4

*i.e.*

In [4]:
cut -d- -f4 NASA_access_log_Jul95 | head

0400] "GET /history/apollo/ HTTP/1.0" 200 6245
0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
0400] "GET /shuttle/missions/sts
0400] "GET /shuttle/countdown/liftoff.html HTTP/1.0" 304 0
0400] "GET /shuttle/missions/sts
0400] "GET /images/NASA
0400] "GET /shuttle/countdown/video/livevideo.gif HTTP/1.0" 200 0
0400] "GET /shuttle/countdown/countdown.html HTTP/1.0" 200 3985
0400] "GET /shuttle/countdown/ HTTP/1.0" 200 3985
0400] "GET / HTTP/1.0" 200 7074


If we use space as our delimiter:
    
    199.72.81.55 - - [01/Jul/1995:00:00:01 -0400] "GET /history/apollo/ HTTP/1.0" 200 6245
    1            2 3 4                     5      6    7                8         9   10

In [5]:
cut -d' ' -f9 NASA_access_log_Jul95 | head

200
200
200
304
200
304
200
200
200
200


In [29]:
cut -d' ' -f9 NASA_access_log_Jul95 | head | grep '[[:digit:]]+'

In [59]:
tail NASA_access_log_Jul95

slipper12055.iaccess.za - - [28/Jul/1995:13:32:22 -0400] "GET /images/USA-logosmall.gif HTTP/1.0" 200 234
163.205.53.14 - - [28/Jul/1995:13:32:22 -0400] "GET /shuttle/technology/images/srb_mod_compare_1-small.gif HTTP/1.0" 200 36902
maynard.isi.uconn.edu - - [28/Jul/1995:13:32:22 -0400] "GET /images/shuttle-patch-logo.gif HTTP/1.0" 200 891
163.205.53.14 - - [28/Jul/1995:13:32:22 -0400] "GET /shuttle/technology/images/srb_mod_compare_3-small.gif HTTP/1.0" 200 55666
163.205.53.14 - - [28/Jul/1995:13:32:22 -0400] "GET /shuttle/technology/images/srb_mod_compare_6-small.gif HTTP/1.0" 200 28219
163.205.53.14 - - [28/Jul/1995:13:32:23 -0400] "GET /images/KSC-logosmall.gif HTTP/1.0" 200 1204
tiger2.ocs.lsu.edu - - [28/Jul/1995:13:32:23 -0400] "GET /shuttle/missions/missions.html HTTP/1.0" 200 8677
199.0.2.27 - - [28/Jul/1995:13:32:23 -0400] "GET /images/ksclogo-medium.gif HTTP/1.0" 200 5866
tornado.umd.edu - - [28/Jul/1995:13:32:25 -0400] "GET /shuttle/missions/sts-74/sts-74-patch-small.gif HT

In [30]:
head -n10000 NASA_access_log_Jul95 | cut -d' ' -f9 | sort | head


10099
11853
1204
1204
17083
200
200
200
200


In [46]:
head -n3950 NASA_access_log_Jul95 | cut -d' ' -f9 | sort | head

11853
200
200
200
200
200
200
200
200
200


In [23]:
cut -d' ' -f9 NASA_access_log_Jul95 | sort | uniq -c | head

cut: NASA_access_log_Jul95: Illegal byte sequence
  77 
   1 -
   1 0
   1 10099
   1 104916
   2 1081049
   1 11175
   1 1121554
   1 11326
   1 11473


Lab
------------------------------------------------
### Together let's
1. `ssh` into our linux box
2. `vi NASA_access_log_Jul95`
3. escape `vi` with `:q!`
2. `vi NASA_access_log_Jul95` again
5. find 400 errors with `/`
6. find 500 errors with `/50[0123456789] `

### On your own:
1. 

In [17]:
grep 'error' NASA_access_log_Jul95 | rev | awk '{print $2}' | head

002
002
002
002
002
002
404
002
002
002


In [15]:
grep 'error' NASA_access_log_Jul95 | head

error170.com - - [07/Jul/1995:01:56:39 -0400] "GET /software/winvn/winvn.html HTTP/1.0" 200 9867
error170.com - - [07/Jul/1995:01:56:42 -0400] "GET /software/winvn/winvn.gif HTTP/1.0" 200 25218
error170.com - - [07/Jul/1995:01:56:42 -0400] "GET /images/construct.gif HTTP/1.0" 200 1414
error170.com - - [07/Jul/1995:01:56:42 -0400] "GET /software/winvn/bluemarb.gif HTTP/1.0" 200 4441
error170.com - - [07/Jul/1995:01:56:43 -0400] "GET /software/winvn/wvsmall.gif HTTP/1.0" 200 13372
error170.com - - [07/Jul/1995:01:56:45 -0400] "GET /images/KSC-logosmall.gif HTTP/1.0" 200 1204
titan02f - - [18/Jul/1995:15:15:39 -0400] "GET /Harvest/cgi-bin/BrokerQuery.pl.cgi?query=oms&host=titan04.ksc.nasa.gov%3A8501&caseflag=on&wordflag=on&errorflag=0&opaqueflag=on&descflag=on&verbose=on&maxresultflag=50 HTTP/1.0" 404 -
titan02f - - [18/Jul/1995:15:18:39 -0400] "GET /~downs/cgi-bin/BrokerQuery.pl.cgi?query=oms&host=titan04.ksc.nasa.gov%3A8501&caseflag=on&wordflag=on&errorflag=0&opaqueflag=on&descflag=on&v

In [33]:
cat NASA_access_log_Jul95 | rev  | awk '{print $2}' | rev | sort | uniq -c

rev: stdin: Illegal byte sequence
   1 "GET
107080 200
4697 302
6577 304
   5 403
 569 404


In [52]:
head -n118928  NASA_access_log_Jul95 | rev | awk '{print $2}' | rev | sort| grep -v 200 | uniq -c 

4697 302
6577 304
   5 403
 569 404


In [53]:
head -n118928  NASA_access_log_Jul95 | rev | awk '{print $2}' | rev | sort | uniq -c 

107080 200
4697 302
6577 304
   5 403
 569 404
